FIAT/macro.py

```diff
@@ -0,0 +1,186 @@
+    foo = numpy.linalg.solve(mat, b)
+    if result is None:
+        return numpy.copy(foo.T)
+    else:
+        result[:, :] = foo[:, :].T
+    return result
+
+
+def facet_support(facet_coords, tol=1.e-12):
+    # facet_coords is an iterable of tuples (barycentric coordinates)
+    # return vertex ids where some x is nonzero
+    return tuple(sorted(set(i for x in facet_coords for (i, xi) in enumerate(x) if abs(xi) > tol)))
+
+
+def invert_cell_topology(T):
+    return {dim: {T[dim][entity]: entity for entity in T[dim]} for dim in T}
+
+
+class SplitSimplicialComplex(SimplicialComplex):
+    """Abstract class to implement a split on a Simplex
+    """
+
+    def __init__(self, ref_el, splits=1, variant=None):
+        self.parent = ref_el
+        vertices, topology = self.split_topology(ref_el, splits=splits, variant=variant)
+        super(SplitSimplicialComplex, self).__init__(ref_el.shape, vertices, topology)
+
+    def split_topology(self, ref_el):
+        raise NotImplementedError
+
+    def get_child_to_parent(self):
+        bary = xy_to_bary(numpy.asarray(self.parent.get_vertices()),
+                          numpy.asarray(self.get_vertices()))
+        top = self.get_topology()
+        parent_inv_top = invert_cell_topology(self.parent.get_topology())
+        child_to_parent = {}
+        for dim in top:
+            child_to_parent[dim] = {}
+            for entity in top[dim]:
+                facet_ids = top[dim][entity]
+                facet_coords = bary[list(facet_ids), :]
+                parent_verts = facet_support(facet_coords)
+                parent_dim = len(parent_verts) - 1
+                parent_entity = parent_inv_top[parent_dim][parent_verts]
+                child_to_parent[dim][entity] = (parent_dim, parent_entity)
+        return child_to_parent
+
+    def get_cell_connectivity(self):
+        """Connectitivity from cell in a complex to globally number and
+        respects the entity numbering on the reference cell.
+        """
+        sd = self.get_spatial_dimension()
+        top = self.get_topology()
+        inv_top = invert_cell_topology(top)
+        parent_top = self.parent.get_topology()
+        connectivity = {cell: {dim: [] for dim in top} for cell in top[sd]}
+        for cell in top[sd]:
+            cell_verts = top[sd][cell]
+            for dim in top:
+                for entity in parent_top[dim]:
+                    ref_verts = parent_top[dim][entity]
+                    global_verts = tuple(cell_verts[v] for v in ref_verts)
+                    connectivity[cell][dim].append(inv_top[dim][global_verts])
+        return connectivity
+
+    def construct_subelement(self, dimension):
+        """Constructs the reference element of a cell subentity
+        specified by subelement dimension.
+
+        :arg dimension: subentity dimension (integer)
+        """
+        return self.parent.construct_subelement(dimension)
+
+
+class AlfeldSplit(SplitSimplicialComplex):
```

Comment:
Documentation: takes a simplex and produces the simplicial complex obtained by connecting vertices to barycenter.  Commentary on the ordering/topology may be wanted as well.